strkey/muxed_account.go

```diff
@@ -0,0 +1,99 @@
+package strkey
+
+import (
+	"bytes"
+	"fmt"
+
+	xdr "github.com/stellar/go-xdr/xdr3"
+	"github.com/stellar/go/support/errors"
+)
+
+type MuxedAccount struct {
+	id      uint64
+	ed25519 [32]byte
+}
+
+// SetID populates the muxed account ID.
+func (m *MuxedAccount) SetID(id uint64) {
+	m.id = id
+}
+
+// SetAddress populates the muxed account ed25519 address.
+func (m *MuxedAccount) SetAddress(address string) error {
+	if !IsValidEd25519PublicKey(address) {
+		return errors.New("invalid ed25519 public key")
+	}
+
+	raw, err := Decode(VersionByteAccountID, address)
+	if err != nil {
+		return errors.Wrap(err, "decoding ed25519 address")
+	}
+	if len(raw) != 32 {
+		return fmt.Errorf("invalid binary length: %d", len(raw))
+	}
+
+	copy(m.ed25519[:], raw)
+
+	return nil
+}
+
+// ID returns the muxed account id according with the SEP-23 definition for
+// multiplexed accounts.
+func (m *MuxedAccount) ID() uint64 {
+	return m.id
+}
+
+// Address returns the muxed account G-address according with the SEP-23
+// definition for multiplexed accounts.
+func (m *MuxedAccount) Address() (string, error) {
+	if m.ed25519 == [32]byte{} {
+		return "", errors.New("muxed account has no ed25519 key")
+	}
+
+	return Encode(VersionByteAccountID, m.ed25519[:])
+}
+
+// MuxedAddress returns the muxed account M-address according with the SEP-23
+// definition for multiplexed accounts.
+func (m *MuxedAccount) MuxedAddress() (string, error) {
+	if m.ed25519 == [32]byte{} {
+		return "", errors.New("muxed account has no ed25519 key")
+	}
+
+	b := new(bytes.Buffer)
+	_, err := xdr.Marshal(b, m.id)
+	if err != nil {
+		return "", errors.Wrap(err, "marshaling muxed address id")
+	}
+
+	raw := make([]byte, 0, 40)
+	raw = append(raw, m.ed25519[:]...)
+	raw = append(raw, b.Bytes()...)
+
+	return Encode(VersionByteMuxedAccount, raw)
+}
+
+// ParseMuxedAccount receives a muxed account M-address and parses it into a
+// MuxedAccount object containing an ed25519 address and an id.
+func ParseMuxedAccount(mAddress string) (*MuxedAccount, error) {
```

Comment:
❗ Looking at the existing strkey API, I think we should give this another name.

<img width="449" alt="Screen Shot 2021-11-08 at 11 13 15 AM" src="https://user-images.githubusercontent.com/351529/140803410-0d13cb1f-3e42-48c8-a1c4-c3e6cdd2f062.png">

It looks like this function is the same as `Decode`, except the versionbyte is known, and therefore we want to return the muxed account type. So, let's use the same name.

```suggestion
func DecodeMuxedAccount(address string) (*MuxedAccount, error) {
```

Reply:
https://github.com/stellar/go/pull/4053/commits/7638e65839fd33cc99f1e9b861c5b5ee2901e38f